tests/runtime/model_management/test_model_manager.py

```diff
@@ -773,6 +774,164 @@ def test_lazy_load_ephemeral_model():
+                    "local_models_dir": cache_dir,
+                    "lazy_load_local_models": True,
+                    "lazy_load_poll_period_seconds": 0,
+                },
+            },
+            "merge",
+        ) as managers:
+            manager = managers[0]
+            with pytest.raises(CaikitRuntimeException) as excinfo:
+                manager.deploy_model("bad-model", {invalid_fname: b"asdf"})
+                assert excinfo.value.status_code == grpc.StatusCode.INVALID_ARGUMENT
+
+
+def test_deploy_with_nested_files(deploy_good_model_files):
+    """Make sure models with nested directories can be deployed"""
+    with TemporaryDirectory() as cache_dir:
+        with non_singleton_model_managers(
+            1,
+            {
+                "runtime": {
+                    "local_models_dir": cache_dir,
+                    "lazy_load_local_models": True,
+                    "lazy_load_poll_period_seconds": 0,
+                },
+            },
+            "merge",
+        ) as managers:
+            manager = managers[0]
+            model_name = "my-model"
+
+            # Read the model files and deploy
+            nested_dir = os.path.join("nested", "twice")
+            nested_fname = "foo.txt"
+            deploy_good_model_files[os.path.join(nested_dir, nested_fname)] = b"foo"
+            loaded_model = manager.deploy_model(
+                model_name, deploy_good_model_files, wait=True
+            )
+            assert loaded_model
+
+            # Make sure the nested file structure was set up correctly
+            local_nested_dir = os.path.join(cache_dir, model_name, nested_dir)
+            assert os.path.isdir(local_nested_dir)
+            assert os.path.exists(os.path.join(local_nested_dir, nested_fname))
+
+
+def test_deploy_invalid_permissions(deploy_good_model_files):
+    """Make sure that an error is raised if attempting to deploy when writing to
+    local_models_dir is denied
+    """
+    with TemporaryDirectory() as cache_dir:
+        local_models_dir = os.path.join(cache_dir, "local_models")
+        os.makedirs(local_models_dir)
+        os.chmod(local_models_dir, 0o600)
+        try:
+            with non_singleton_model_managers(
+                1,
+                {
+                    "runtime": {
+                        "local_models_dir": local_models_dir,
+                        "lazy_load_local_models": True,
+                        "lazy_load_poll_period_seconds": 0,
+                    },
+                },
+                "merge",
+            ) as managers:
+                manager = managers[0]
+                model_name = "my-model"
+
+                # Make sure the deploy fails with a permission error
+                with pytest.raises(CaikitRuntimeException) as excinfo:
+                    manager.deploy_model(model_name, deploy_good_model_files, wait=True)
+                    assert excinfo.status_code == grpc.StatusCode.PERMISSION_DENIED
```

Comment:
I think PERMISSION_DENIED implies that the caller doesn't have permission to perform the action, but this error case is that the user id running the server itself does not have permission to write the files. Is this a `FAILED_PRECONDITION`?


https://grpc.github.io/grpc/core/md_doc_statuscodes.html
>  The operation was rejected because the system is not in a state required for the operation's execution. For example, the directory to be deleted is non-empty, an rmdir operation is applied to a non-directory, etc. Service implementors can use the following guidelines to decide between FAILED_PRECONDITION, ABORTED, and UNAVAILABLE: (a) Use UNAVAILABLE if the client can retry just the failing call. (b) Use ABORTED if the client should retry at a higher level (e.g., when a client-specified test-and-set fails, indicating the client should restart a read-modify-write sequence). (c) Use FAILED_PRECONDITION if the client should not retry until the system state has been explicitly fixed. E.g., if an "rmdir" fails because the directory is non-empty, FAILED_PRECONDITION should be returned since the client should not retry unless the files are deleted from the directory.

Reply:
Ah, very good call. Will switch to `FAILED_PRECONDITION`